### ISSUE TITLE
`no-magic-number` issues

### ISSUE THREAD

**Issue by ilyavolodin** [eslint/eslint#4616]
I was working on fix for #4370 and I got very confused by the logic of this rule. Fix itself is easy to add, but I don't understand how some of the unittests are passing and I also don't understand what `detectObjects` supposed to do. `detectObjects` is very poorly tested and doesn't have a single valid test case. It's also clearly incorrect (as it will ignore any number who's parent has a type of `AssignmentExpression` which is the whole point of this rule). Can somebody explain what `detectObjects` supposed to do and why are the following unit tests passing:
https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-magic-numbers.js#L34
https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-magic-numbers.js#L38
https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-magic-numbers.js#L45
https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-magic-numbers.js#L60
https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-magic-numbers.js#L63

**Comment by IanVS** [comment]
I'm going to take a wild guess here about why those tests are passing, but I have no real insider knowledge here.

I think `const foo = 42;` has to pass because it is assigning to a `const`, presumably in order to avoid using the `42` later on in the code (thus avoiding a magic number).

**Comment by ilyavolodin** [comment]
Ok, you are absolutely right about those tests. I lost the forest behind the trees, and it makes sense that assigning value to a variable should pass. So those unittests are passing correctly. Based on the code, however, I'm not sure why is `var min, max, mean; min = 1; max = 10; mean = 4;` passing, since the exception is made for `VariableDeclarator` and not for `AssignmentExpression`.
As far as I understand from the code `detectObjects` supposed to not flag following expressions `var a = {}; a.9 = 'test'` and `var a = { 9: 'test' }`. First one, however, is illegal syntax, and will not even parse, second one should be so incredibly rare, that I'm not sure what would be the point of creating a separate option for it. In the current implementation if `detectObjects` is on, it will also start flagging any magic numbers in `AssignmentExpression`, which is very strange, since it will stop flagging them even if they are outside of the object declaration.

**Comment by ilyavolodin** [comment]
@kombucha any insights?

**Comment by nzakas** [comment]
These seem like they shouldn't be valid:

```js
var min, max, mean; min = 1; max = 10; mean = 4;

var a = {}; a[9] = 'test';
```

The intent of `detectObjects` was to toggle allowing this:

```js
var foo = {
    bar: 5
};
```

I agree, this is confusing. We should figure out a way to clean this up.

**Comment by nzakas** [comment]
@ilyavolodin any followup on this?

**Comment by ilyavolodin** [comment]
I'm not really sure what to do about those issues. I didn't participate in the original discussion and don't know the reasoning behind some of the decisions for this rule. Since we are not getting any response from the author, I think we basically have to re-write this rule and clean-up documentation and tests.

**Comment by alberto** [comment]
What do we want to do here?

```
var a = {}; a[9] = 'test';
```
This is not a valid test case, @ilyavolodin I think you might have gotten the behaviour backwards. @nzakas example is the correct one of what this is supposed to flag. 

So, I think the only issue is if we want this rule to flag or not:

```
var min, max, mean; min = 1; max = 10; mean = 4;
```

**Comment by nzakas** [comment]
I think we do.

**Comment by alberto** [comment]
Accepting this as a bug, then.

**Comment by not-an-aardvark** [comment]
Working on this.
